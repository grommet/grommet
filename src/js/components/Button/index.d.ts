import { string } from 'prop-types';
import * as React from 'react';
import {
  A11yTitleType,
  AlignSelfType,
  BackgroundType,
  ColorType,
  FillType,
  GapType,
  GridAreaType,
  HoverIndicatorType,
  MarginType,
  Omit,
  PadType,
  PolymorphicType,
} from '../../utils';

import { TipProps } from '../Tip';

export interface ButtonProps {
  a11yTitle?: A11yTitleType;
  alignSelf?: AlignSelfType;
  badge?:
    | boolean
    | number
    | {
        background?: BackgroundType;
        max?: number;
        value?: boolean | number;
      }
    | JSX.Element;
  children?:
    | React.ReactNode
    | (({
        disabled,
        hover,
        focus,
      }: {
        disabled: boolean;
        hover: boolean;
        focus: boolean;
      }) => React.ReactNode);
  busy?: boolean;
  gridArea?: GridAreaType;
  margin?: MarginType;
  active?: boolean;
  color?: ColorType;
  disabled?: boolean;
  fill?: FillType;
  focusIndicator?: boolean;
  gap?: GapType;
  hoverIndicator?: HoverIndicatorType;
  href?: string;
  justify?:
    | 'around'
    | 'between'
    | 'center'
    | 'end'
    | 'evenly'
    | 'start'
    | 'stretch';
  target?: '_self' | '_blank' | '_parent' | '_top' | string;
  icon?: JSX.Element;
  kind?: string;
  label?: React.ReactNode;
  messages?: {
    busy?: string;
    success?: string;
  };
  pad?: PadType;
  plain?: boolean;
  primary?: boolean;
  reverse?: boolean;
  secondary?: boolean;
  size?: 'small' | 'medium' | 'large' | string;
  success?: boolean;
  tip?: TipProps | string;
  type?: 'button' | 'reset' | 'submit';
  as?: PolymorphicType;
}

type anchorType = Omit<JSX.IntrinsicElements['a'], 'color'>;
type buttonType = Omit<JSX.IntrinsicElements['button'], 'color'>;
type extendType = anchorType & buttonType;

export interface ButtonExtendedProps
  extends ButtonProps,
    Omit<extendType, 'target' | 'children'> {}

// Keep type alias for backwards compatibility.
export type ButtonType = ButtonProps & extendType;

declare const Button: React.FC<ButtonExtendedProps>;

export { Button };
