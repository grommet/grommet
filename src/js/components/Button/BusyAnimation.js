import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import { Checkmark } from 'grommet-icons/icons/Checkmark';
import { Box } from '../Box';
import { styledComponentsConfig } from '../../utils/styles';

const bounceDelay = keyframes`
  0%, 80%, 100% { 
    transform: scale(0.4);
  } 40% { 
    transform: scale(0.8);
  }
`;

const bounceDelayRule = css`
  animation: ${bounceDelay} 1.4s infinite ease-in-out both;
`;

/* When button is small size we need half the dot size to fit properly */
const Dot = styled(Box)`
  background-color: currentColor;
  width: ${(props) => (props.size.size === 'small' ? '4px' : '8px')};
  height: ${(props) => (props.size.size === 'small' ? '4px' : '8px')};
  border-radius: 100%;
  display: inline-block;
  ${bounceDelayRule}
  ${(props) => props.delay && `animation-delay: ${props.delay};`}
`;

export const EllipsisAnimation = (size) => (
  <Box
    style={{ position: 'absolute' }}
    fill
    alignContent="center"
    justify="center"
  >
    <Box alignSelf="center" direction="row" gap="small">
      {/* A negative delay starts the animation sooner. The first dot
      should begin animating before the second and so on. */}
      <Dot size={size} delay="-0.32s" />
      <Dot size={size} delay="-0.16s" />
      <Dot size={size} />
    </Box>
  </Box>
);

const grow = keyframes`
  0% {
    opacity: 0;
    transform: scale(.3);
  }
  20% {
    opacity: 1;
    transform: scale(1.15);
  }
  30% { transform: scale(.9); }
  45% { transform: scale(1.05); }
  55% { transform: scale(1); }
  100% { transform: scale(1); }  
`;

export const GrowCheckmark = styled(Checkmark)`
  position: absolute;
  align-self: center;
  animation: ${grow} 0.9s ease-in-out;
`;

export const StyledBusyContents = styled.div.withConfig(styledComponentsConfig)`
  opacity: ${(props) => (props.animating ? 0 : 1)};}
`;
