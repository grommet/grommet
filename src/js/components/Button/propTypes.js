import PropTypes from 'prop-types';
import {
  backgroundPropType,
  genericProps,
  colorPropType,
  hoverIndicatorPropType,
  padPropType,
} from '../../utils/general-prop-types';

let PropType = {};
if (process.env.NODE_ENV !== 'production') {
  PropType = {
    ...genericProps,
    children: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.object,
      PropTypes.node,
    ]),
    active: PropTypes.bool,
    as: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
      PropTypes.elementType,
    ]),
    badge: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.element,
      PropTypes.number,
      PropTypes.shape({
        background: backgroundPropType,
        max: PropTypes.number,
        value: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
      }),
    ]),
    busy: PropTypes.bool,
    color: colorPropType,
    disabled: PropTypes.bool,
    fill: PropTypes.oneOfType([
      PropTypes.oneOf(['horizontal', 'vertical']),
      PropTypes.bool,
    ]),
    focusIndicator: PropTypes.bool,
    gap: PropTypes.oneOfType([
      PropTypes.oneOf([
        'none',
        'xxsmall',
        'xsmall',
        'small',
        'medium',
        'large',
        'xlarge',
      ]),
      PropTypes.string,
    ]),
    hoverIndicator: hoverIndicatorPropType,
    href: PropTypes.string,
    icon: PropTypes.element,
    justify: PropTypes.oneOf([
      'around',
      'between',
      'center',
      'end',
      'evenly',
      'start',
      'stretch',
    ]),
    label: PropTypes.node,
    messages: PropTypes.shape({
      busy: PropTypes.string,
      success: PropTypes.string,
    }),
    onClick: PropTypes.func,
    pad: padPropType,
    plain: PropTypes.bool,
    primary: PropTypes.bool,
    reverse: PropTypes.bool,
    secondary: PropTypes.bool,
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['small', 'medium', 'large']),
      PropTypes.string,
    ]),
    success: PropTypes.bool,
    target: PropTypes.oneOfType([
      PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
      PropTypes.string,
    ]),
    tip: PropTypes.oneOfType([
      PropTypes.shape({
        content: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
        dropProps: PropTypes.shape({}),
        plain: PropTypes.bool,
      }),
      PropTypes.string,
    ]),
    type: PropTypes.oneOf(['button', 'reset', 'submit']),
  };
}
export const ButtonPropTypes = PropType;
