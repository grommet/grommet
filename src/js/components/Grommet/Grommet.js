import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  deviceResponsive,
  getBreakpoint,
  normalizeColor,
  useForwardedRef,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';
import { RootsContext } from '../../contexts/RootsContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { format, MessageContext } from '../../contexts/MessageContext';
import defaultMessages from '../../languages/default.json';
import { GrommetPropTypes } from './propTypes';
import { AnalyticsProvider } from '../../contexts/AnalyticsContext';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const defaultOptions = {};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    options = defaultOptions,
    messages: messagesProp,
    onAnalytics,
    ...rest
  } = props;
  const { background, dir, themeMode, userAgent } = props;
  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (
      themeProp &&
      themeProp.menu &&
      themeProp.menu.drop &&
      themeProp.menu.drop.align
    ) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';

    if (
      themeMode === 'auto' &&
      typeof window !== 'undefined' &&
      window.matchMedia &&
      window.matchMedia('(prefers-color-scheme: dark)').matches
    ) {
      nextTheme.dark = true;
    }

    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  const messages = useMemo(() => {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    const nextMessages = deepMerge(
      defaultMessages,
      messagesProp?.messages || {},
    );
    return {
      messages: nextMessages,
      format: (opts) => {
        const message = messagesProp?.format && messagesProp.format(opts);
        return typeof message !== 'undefined'
          ? message
          : format(opts, nextMessages);
      },
    };
  }, [messagesProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  const grommetRef = useForwardedRef(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  const roots = useRef([]);
  useEffect(() => {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  const rootsContextValue = useMemo(() => ({ roots }), []);

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <RootsContext.Provider value={rootsContextValue}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <OptionsContext.Provider value={options}>
              <MessageContext.Provider value={messages}>
                <AnalyticsProvider onAnalytics={onAnalytics}>
                  <StyledGrommet full={full} {...rest} ref={grommetRef}>
                    {children}
                  </StyledGrommet>
                  {full && <FullGlobalStyle />}
                </AnalyticsProvider>
              </MessageContext.Provider>
            </OptionsContext.Provider>
          </ContainerTargetContext.Provider>
        </RootsContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';
Grommet.propTypes = GrommetPropTypes;

export { Grommet };
