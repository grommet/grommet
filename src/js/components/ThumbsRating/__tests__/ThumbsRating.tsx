import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import 'jest-styled-components';
import 'jest-axe/extend-expect';
import 'regenerator-runtime/runtime';
import '@testing-library/jest-dom';

import { Button } from '../../Button';
import { Grommet } from '../../Grommet';
import { Form } from '../../Form';
import { FormField } from '../../FormField';
import { ThumbsRating } from '..';
import { Add, AddCircle, Subtract, SubtractCircle } from 'grommet-icons';

describe('ThumbsRating', () => {
  test('StarRating should have no accessibility violations', async () => {
    const { container } = render(
      <Grommet>
        <ThumbsRating name="test" />
      </Grommet>,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  test('StarRating is present', () => {
    const { container, getByTestId } = render(
      <Grommet>
        <ThumbsRating name="test" data-testid="ThumbsRating-icon" />
      </Grommet>,
    );

    expect(getByTestId('ThumbsRating-icon')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('StarRating renders outside grommet wrapper', () => {
    const { container } = render(<ThumbsRating name="test" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('value for thumbs', async () => {
    const onSubmit = jest.fn();
    const user = userEvent.setup();
    const Test = () => (
      <Form onSubmit={({ value }) => onSubmit({ value })}>
        <FormField label="test-starRating">
          <ThumbsRating name="test" />
        </FormField>
        <Button label="submit" type="submit" />
      </Form>
    );
    const { container, getByText } = render(
      <Grommet>
        <Test />
      </Grommet>,
    );
    expect(container.firstChild).toMatchSnapshot();
    await user.tab();
    await user.keyboard('{arrowdown}');
    fireEvent.click(getByText('submit'));
    expect(onSubmit).toHaveBeenCalledWith(
      expect.objectContaining({
        value: { test: 'dislike' },
      }),
    );
  });

  test('should use icons from theme', async () => {
    const user = userEvent.setup();
    render(
      <Grommet
        theme={{
          thumbsRating: {
            icons: {
              like: Add,
              likeSelected: AddCircle,
              dislike: Subtract,
              dislikeSelected: SubtractCircle,
            },
          },
        }}
      >
        <ThumbsRating name="test" data-testid="ThumbsRating-icon" />
      </Grommet>,
    );

    expect(screen.getByLabelText('Add')).toBeInTheDocument();
    expect(screen.getByLabelText('Subtract')).toBeInTheDocument();

    await user.tab();
    await user.keyboard('{arrowdown}');
    expect(screen.getByLabelText('SubtractCircle')).toBeInTheDocument();
    await user.keyboard('{arrowdown}');
    expect(screen.getByLabelText('AddCircle')).toBeInTheDocument();
  });
});
