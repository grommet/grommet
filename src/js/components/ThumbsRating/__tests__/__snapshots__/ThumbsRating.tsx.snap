// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbsRating StarRating is present 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c7 {
  flex: 0 0 auto;
  align-self: stretch;
  width: 12px;
}

.c6 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c6 g {
  fill: inherit;
  stroke: inherit;
}

.c6 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c6 *[stroke*='#'],
.c6 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c6 *[fill-rule],
.c6 *[FILL-RULE],
.c6 *[fill*='#'],
.c6 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c2 {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
  width: fit-content;
  cursor: pointer;
}

.c2:hover input:not([disabled])+div,
.c2:hover input:not([disabled])+span {
  border-color: #000000;
}

.c4 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width: 768px) {
  .c7 {
    width: 6px;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="ThumbsRating-icon"
      role="radiogroup"
    >
      <label
        class="c2"
        for="like"
      >
        <div
          class="c3 "
        >
          <input
            class="c4"
            id="like"
            name="test"
            tabindex="0"
            type="radio"
            value="like"
          />
          <div
            class="c5 "
          >
            <svg
              aria-label="Like"
              class="c6"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </label>
      <div
        class="c7"
      />
      <label
        class="c2"
        for="dislike"
      >
        <div
          class="c3 "
        >
          <input
            class="c4"
            id="dislike"
            name="test"
            tabindex="-1"
            type="radio"
            value="dislike"
          />
          <div
            class="c5 "
          >
            <svg
              aria-label="Dislike"
              class="c6"
              viewBox="0 0 24 24"
            >
              <path
                d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`ThumbsRating StarRating renders outside grommet wrapper 1`] = `
.c0 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
}

.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c6 {
  flex: 0 0 auto;
  align-self: stretch;
  width: 12px;
}

.c5 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c5 g {
  fill: inherit;
  stroke: inherit;
}

.c5 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c5 *[stroke*='#'],
.c5 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c5 *[fill-rule],
.c5 *[FILL-RULE],
.c5 *[fill*='#'],
.c5 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
  width: fit-content;
  cursor: pointer;
}

.c1:hover input:not([disabled])+div,
.c1:hover input:not([disabled])+span {
  border-color: #000000;
}

.c3 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width: 768px) {
  .c6 {
    width: 6px;
  }
}

<div
  class="c0"
  role="radiogroup"
>
  <label
    class="c1"
    for="like"
  >
    <div
      class="c2 "
    >
      <input
        class="c3"
        id="like"
        name="test"
        tabindex="0"
        type="radio"
        value="like"
      />
      <div
        class="c4 "
      >
        <svg
          aria-label="Like"
          class="c5"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </label>
  <div
    class="c6"
  />
  <label
    class="c1"
    for="dislike"
  >
    <div
      class="c2 "
    >
      <input
        class="c3"
        id="dislike"
        name="test"
        tabindex="-1"
        type="radio"
        value="dislike"
      />
      <div
        class="c4 "
      >
        <svg
          aria-label="Dislike"
          class="c5"
          viewBox="0 0 24 24"
        >
          <path
            d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </label>
</div>
`;

exports[`ThumbsRating StarRating should have no accessibility violations 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c7 {
  flex: 0 0 auto;
  align-self: stretch;
  width: 12px;
}

.c6 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c6 g {
  fill: inherit;
  stroke: inherit;
}

.c6 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c6 *[stroke*='#'],
.c6 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c6 *[fill-rule],
.c6 *[FILL-RULE],
.c6 *[fill*='#'],
.c6 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c2 {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
  width: fit-content;
  cursor: pointer;
}

.c2:hover input:not([disabled])+div,
.c2:hover input:not([disabled])+span {
  border-color: #000000;
}

.c4 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width: 768px) {
  .c7 {
    width: 6px;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      role="radiogroup"
    >
      <label
        class="c2"
        for="like"
      >
        <div
          class="c3 "
        >
          <input
            class="c4"
            id="like"
            name="test"
            tabindex="0"
            type="radio"
            value="like"
          />
          <div
            class="c5 "
          >
            <svg
              aria-label="Like"
              class="c6"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </label>
      <div
        class="c7"
      />
      <label
        class="c2"
        for="dislike"
      >
        <div
          class="c3 "
        >
          <input
            class="c4"
            id="dislike"
            name="test"
            tabindex="-1"
            type="radio"
            value="dislike"
          />
          <div
            class="c5 "
          >
            <svg
              aria-label="Dislike"
              class="c6"
              viewBox="0 0 24 24"
            >
              <path
                d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`ThumbsRating value for thumbs 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-bottom: 12px;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c8 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c10 {
  flex: 0 0 auto;
  align-self: stretch;
  width: 12px;
}

.c9 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c9 g {
  fill: inherit;
  stroke: inherit;
}

.c9 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c9 *[stroke*='#'],
.c9 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c9 *[fill-rule],
.c9 *[FILL-RULE],
.c9 *[fill*='#'],
.c9 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c2 {
  margin-left: 12px;
  margin-right: 12px;
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c11 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c11:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c11:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c11:focus >circle,
.c11:focus >ellipse,
.c11:focus >line,
.c11:focus >path,
.c11:focus >polygon,
.c11:focus >polyline,
.c11:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c11:focus ::-moz-focus-inner {
  border: 0;
}

.c11:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c11:focus:not(:focus-visible) >circle,
.c11:focus:not(:focus-visible) >ellipse,
.c11:focus:not(:focus-visible) >line,
.c11:focus:not(:focus-visible) >path,
.c11:focus:not(:focus-visible) >polygon,
.c11:focus:not(:focus-visible) >polyline,
.c11:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c11:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c5 {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
  width: fit-content;
  cursor: pointer;
}

.c5:hover input:not([disabled])+div,
.c5:hover input:not([disabled])+span {
  border-color: #000000;
}

.c7 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width: 768px) {
  .c1 {
    margin-bottom: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c3 {
    border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  }
}

@media only screen and (max-width: 768px) {
  .c10 {
    width: 6px;
  }
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 "
    >
      <label
        class="c2"
      >
        test-starRating
      </label>
      <div
        class="c3 "
      >
        <div
          class="c4"
          role="radiogroup"
        >
          <label
            class="c5"
            for="like"
          >
            <div
              class="c6 "
            >
              <input
                class="c7"
                id="like"
                name="test"
                tabindex="0"
                type="radio"
                value="like"
              />
              <div
                class="c8 "
              >
                <svg
                  aria-label="Like"
                  class="c9"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </label>
          <div
            class="c10"
          />
          <label
            class="c5"
            for="dislike"
          >
            <div
              class="c6 "
            >
              <input
                class="c7"
                id="dislike"
                name="test"
                tabindex="-1"
                type="radio"
                value="dislike"
              />
              <div
                class="c8 "
              >
                <svg
                  aria-label="Dislike"
                  class="c9"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </label>
        </div>
      </div>
    </div>
    <button
      class="c11"
      type="submit"
    >
      submit
    </button>
  </form>
</div>
`;
